[Non-Root] Add support for running without root permissions on Linux

blakerouse opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

Adds the ability to install the elastic-agent where it runs as its own elastic-agent user on Linux. Limited to Linux at the moment just to keep the PR smaller.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Allows the ability to install an Elastic Agent that doesn't require root access.

**Note:** Root is still required to install the Elastic Agent, just adds support for `--unprivileged` option.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~ (not possible to unit test, would affect the actual system)
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~ (will add changelog when its enabled on all platforms)
- [x] I have added an integration test or an E2E test (TODO)

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [x] Installs and runs
- [x] All files owned by `elastic-agent:elastic-agent`
- [x] Control socket can be used when user is added to `elastic-agent` group

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

```
$ sudo ./elastic-agent install --unprivileged
$ sudo usermod -a -G elastic-agent $USER
$ newgrp elastic-agent
$ elastic-agent status
```

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3598/detail/PR-3598/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3598/detail/PR-3598/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3598/detail/PR-3598/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3598/detail/PR-3598/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-agent_3598.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-10-12T20:41:34.592Z&rangeTo=2023-10-12T21:01:34.592Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=ab4a20ab8822796ad00212e9dfc2348d&transactionId=3ecfa5cc4d804f62)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-10-12T20:51:34.592+0000


* Duration: 19 min 20 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3257  |
| Skipped      | 21 |
| Total        | 3278   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3598/detail/PR-3598/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Go build`
<ul>
<li>Took 3 min 33 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent/pipelines/elastic-agent-mbp/pipelines/PR-3598/runs/1/steps/539/log/?start=0">here</a></li>
<li>Description: <code>mage build</code></l1>
</ul>
  
  
  
  
##### `Go build`
<ul>
<li>Took 2 min 27 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/elastic-agent/pipelines/elastic-agent-mbp/pipelines/PR-3598/runs/1/steps/443/log/?start=0">here</a></li>
<li>Description: <code>mage build</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.78% (`81/82`) | :+1: 
Files | 66.333% (`199/300`) | :-1: -0.67
Classes | 65.288% (`363/556`) | :-1: -0.354
Methods | 52.682% (`1149/2181`) | :-1: -0.291
Lines | 38.382% (`13142/34240`) | :-1: -0.196
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

blakerouse commented:
This PR enables installation and running as non-root. It does not enable upgrading as non-root, there is an issue with the watcher where it tries to make calls to the service runtime to watch the PID. A non-root user doesn't have that ability to perform those actions. A solution is needed for this problem.

blakerouse commented:
Moving in draft as I work on a watcher change required for non-root upgrading to work.

blakerouse commented:
Currently hitting issue with Elastic Agent not showing or using the correct build when testing self-upgrade with local build using `file://` as the `--source-uri`. Working on this PR to fix the issue and provide a better experience using a local file for upgrades.

https://github.com/elastic/elastic-agent/pull/3682

blakerouse commented:
**Note**: I cannot add an upgrade test, until latest SNAPSHOT exists with this change in it. This change requires the version to upgrade to to support non-root.

blakerouse commented on internal/pkg/agent/install/perms.go:
```diff
@@ -0,0 +1,10 @@
+
+package install
+
+// FixPermissions fixes the permissions on the installed system.
+func FixPermissions(topPath string, uid string, gid string) error {
+	return fixPermissions(topPath, uid, gid)
+}
```
Because each OS has a different `fixPermissions`. I preferred for the public function to be for all OS's and the implementation unique per OS.

blakerouse replied:
Changed to remove the extra `FixPermissions` wrapper.

blakerouse commented on internal/pkg/agent/cmd/install.go:
```diff
@@ -21,7 +22,10 @@ import (
-const flagInstallBasePath = "base-path"
+const (
+	flagInstallBasePath = "base-path"
+	flagInstallNonRoot  = "non-root"
```
We want the default to be installation as root as that gives the best experience and allows the ability for Endpoint Security to be used. By changing to `--root` that would mean the default becomes non-root which is not what we want.

We will also not allow the ability user a specific user. It will always be the `elastic-agent` user and the `elastic-agent` group. They can always add a user to `elastic-agent` group to allow them to interact with the Elastic Agent.

cmacknz replied:
Agree, we can't change the default which means we need a negative in the flag like this.

My only other suggestion would be to come up with a name for this option that isn't Linux specific, since root doesn't exist on Windows. It's Administrator there. Perhaps `--unprivileged`? Not sure if that is as clear or if there is a better general term for this.

Or maybe just aliasing this to `--non-admin` works for people who only speak Windows.

blakerouse replied:
I like `--unprivileged`, it works for both. Instead of having multiple flags depending on the OS. If there is an agreement on `--unprivileged` I will change it to that.

nimarezainia replied:
i also agree with `--unprivileged` we need to ensure this doesn't cause a breaking change.

blakerouse replied:
Changed to `--unprivileged`.

blakerouse commented on internal/pkg/agent/application/paths/paths_windows.go:
```diff
@@ -22,6 +22,9 @@ const (
 	ControlSocketPath = `\\.\pipe\elastic-agent-system`
 
+	// ControlSocketNonRootPath is the control socket path used when installed as non-root.
+	ControlSocketNonRootPath = `\\.\pipe\elastic-agent-system`
```
Done.

blakerouse commented on internal/pkg/agent/cmd/run.go:
```diff
@@ -612,7 +612,7 @@ func ensureInstallMarkerPresent() error {
 	// that didn't use an installation marker file (that is, before v8.8.0).
 	// So create the file now.
-	if err := info.CreateInstallMarker(paths.Top()); err != nil {
+	if err := info.CreateInstallMarker(paths.Top(), "", ""); err != nil {
```
Done.

cmacknz commented on testing/integration/install_non_root_test.go:
```diff
@@ -0,0 +1,205 @@
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+)
+
+func TestInstallNonRootWithoutBasePath(t *testing.T) {
```
I like these tests but I'm not sure they are covering enough to ensure we never break the non-root mode. I am worried anytime somebody does anything involving a file they will break the non-root mode by failing to set the permissions properly, or placing it in a directory the elastic-agent user can't access by default.

Key things I can think of that don't look like they are covered here:

1. Upgrades.
2. Enrolling with Fleet
3. Shipping monitoring logs to Fleet.
4. Collecting diagnostics.
5. Collecting system logs and metrics as the system integration would do.

cmacknz replied:
I wonder if it would be better to run every test in non-root mode by default since that is the hardest to get right, and only run the sanity check tests for the agent as root. So the inverse of this approach since if it works as non-root it is highly likely to work as root.

The endpoint security tests would have to be root at least, that plus an upgrade test as root might be enough as a sanity check.

cmacknz replied:
The main difference seems to be the socket path being different and the install operation has to create the user and group, otherwise it seems we are just auditing for incorrect file permissions.

blakerouse replied:
I could do that in a follow up PR. Especially because I have to wait for this to be in the SNAPSHOT build before I can write the upgrade test, which has to be in a follow up PR.

cmacknz replied:
> I could do that in a follow up PR

Just to confirm, making the non-root agent the default agent in tests? Would be tidier as a follow up PR anyway most likely.

blakerouse replied:
I will do a follow up once this change is in the latest snapshot to run as many tests as we can both as root and non-root installations.

cmacknz commented on internal/pkg/agent/install/install.go:
```diff
@@ -44,18 +48,62 @@ func Install(cfgFile, topPath string, pt *progressbar.ProgressBar, streams *cli.
+	username := ""
+	groupName := ""
+	if nonRoot {
+		username = elasticUsername
```
nit: This block is big enough to be its own function.

blakerouse replied:
Okay will refactor it.

cmacknz commented on internal/pkg/agent/install/install.go:
```diff
@@ -301,3 +369,22 @@ func hasAllSSDs(block ghw.BlockInfo) bool {
 	return true
 }
+
+func createSocketDir(uid int, gid int) error {
```
UID and GID are passed around so much it might be worth encapsulating them into a structure. At minimum that would let us initialize them by name to minimize the chance that someone accidentally reverses them.

blakerouse replied:
I like that sugession. Will make that change.

ycombinator replied:
If you're going this struct route (which I'm also ++ for), I'd consider encapsulating the string counterparts in the struct and adding accessor methods for returning the corresponding int values.  You could probably just do the conversion once, as part of the struct's constructor, not every time the accessors are called.  Currently, this string->int conversion is being done in a couple of places so moving this into a single place would be nice.

blakerouse replied:
Changing the the struct route allows me to define different structs between unix and Windows, which made the code a lot cleaner and much easier to understand.

cmacknz commented on internal/pkg/agent/application/upgrade/rollback.go:
```diff
@@ -175,19 +179,21 @@ func restartAgent(ctx context.Context, log *logger.Logger) error {
-			break
+		// Next, try to restart Agent via the service. (only if root)
+		if root {
+			log.Infof("Restarting Agent via service; attempt %d of %d", restartAttempt, maxRestartCount)
```
Do we need to keep this in the root case, if it isn't possible in the root case?

I am worried about upgrade bugs becoming specific to the type of installation, although maybe that is unavoidable.

blakerouse replied:
This was added by @ycombinator. It think it was added to ensure that in the worse case that it would be recoverable. Sadly its just not possible to perform in non-root.

cmacknz commented on internal/pkg/agent/application/paths/paths.go:
```diff
@@ -17,6 +17,9 @@ const (
 	ControlSocketPath = "unix:///run/elastic-agent.sock"
 
+	// ControlSocketNonRootPath is the control socket path used when installed as non-root.
+	ControlSocketNonRootPath = "unix:///run/elastic-agent/elastic-agent.sock"
```
Why does this need a containing directory when not running as root? Because of the permissions on /run?

blakerouse replied:
Correct. The `/run/elastic-agent` directory must be created during install when executing as root, and then the `elastic-agent` user can write to that directory.

blakerouse replied:
Added comment on why it must be in a directory.

cmacknz reviewed (commented):
My primary concern is about keeping this working long term, we may need to adjust our testing strategy. Otherwise the changes look pretty reasonable.

We should additionally add something to our diagnostics to clearly indicate if the agent was installed as root or not so we can isolate problems specific to one mode or the other quickly. We may want to log the user we are running as at startup as well or something.

blakerouse commented on testing/integration/install_non_root_test.go:
```diff
@@ -0,0 +1,205 @@
+
+	topPath := filepath.Join(defaultBasePath, "Elastic", "Agent")
+	err = os.RemoveAll(topPath)
+	require.NoError(t, err, "failed to remove %q. The test requires this path not to exist.")
```
It is actually from the other install test that already exists. They do cleanup and it is checked at the end of the test, but if your running with `KEEP_INSTALLED=true` then it can result in it not being removed. This check is just a pre-check to ensure that its clean before performing the test. I think its best to have it just for sanity check before the test runs.

ycombinator commented on internal/pkg/agent/application/info/state.go:
```diff
@@ -28,11 +31,33 @@ func RunningInstalled() bool {
 	}
 
+	var err error
+	if runtime.GOOS != "windows" {
```
Since `os.Getuid()` and `os.Getgid()` do actually work on Windows, it's not immediately obvious (without the context of this PR) why we need this `if` condition here.  Consider adding an explanatory comment.

ycombinator commented on internal/pkg/agent/install/install.go:
```diff
@@ -23,13 +24,16 @@ import (
 )
 
 // Install installs Elastic Agent persistently on the system including creating and starting its service.
-func Install(cfgFile, topPath string, pt *progressbar.ProgressBar, streams *cli.IOStreams) error {
+func Install(cfgFile, topPath string, nonRoot bool, pt *progressbar.ProgressBar, streams *cli.IOStreams) (string, string, error) {
```
I know you mentioned it in the PR description that you're returning UID and GID as strings for future Windows support.  Consider mentioning that in this function's godoc as well for developers reading this code in the future without the context of this PR.

blakerouse replied:
I was able to change this behavior removing the need for these type of comments.

ycombinator commented:
> Note: Root is still required to install the Elastic Agent, just adds support for --non-root option.

UX suggestion: I think it might be a nice UX touch to clarify this in the help text that's displayed when a user tries to install elastic-agent without sudo.  Today, we display the following help text:

```
$ ./elastic-agent install -n
Error: unable to perform install command, not executed with root permissions
For help, please see our troubleshooting guide at https://www.elastic.co/guide/en/fleet/8.12/fleet-troubleshooting.html
```

I think we could add something here that says if you want to _run_ Elastic Agent as an unprivileged user, you still need to execute the install command as a privileged user but should pass the `--unprivileged` option.

ycombinator commented:
I built Agent from this PR's branch and tried installing it with `--non-root`.  I think there are some issues with the socket creation.

```
$ sudo ./elastic-agent install -n --non-root
Installing in non-interactive mode.
[    ] Done  [16s]
Elastic Agent has been successfully installed.
sudo elastic-agent version
Binary: 8.12.0-SNAPSHOT (build: 28faa54f1f3c0cdd50c280d8f3a5bad6a348046a at 2023-11-08 06:54:37 +0000 UTC)
Daemon: <failed to communicate>
could not get version. failed to communicate with running daemon: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /run/elastic-agent.sock: connect: no such file or directory"
Use --binary-only flag to skip trying to retrieve version from running daemon
```

If I try installing the same build of Agent without the `--non-root` option, socket creation works:

```
$ sudo ./elastic-agent install -n
Installing in non-interactive mode.
[=== ] Done  [16s]
Elastic Agent has been successfully installed.
$ sudo elastic-agent version
Binary: 8.12.0-SNAPSHOT (build: 28faa54f1f3c0cdd50c280d8f3a5bad6a348046a at 2023-11-08 06:54:37 +0000 UTC)
Daemon: 8.12.0-SNAPSHOT (build: 28faa54f1f3c0cdd50c280d8f3a5bad6a348046a at 2023-11-08 06:54:37 +0000 UTC)
```

michalpristas commented on internal/pkg/agent/application/paths/paths_darwin.go:
```diff
@@ -17,6 +17,9 @@ const (
 	ControlSocketPath = "unix:///var/run/elastic-agent.sock"
 
+	// ControlSocketNonRootPath is the control socket path used when installed as non-root.
+	ControlSocketNonRootPath = "unix:///var/run/elastic-agent/elastic-agent.sock"
```
could you add a comment why we need to place it in subdir, i'm sure we'll forget

blakerouse replied:
Added.

michalpristas commented on internal/pkg/agent/cmd/install.go:
```diff
@@ -67,6 +72,12 @@ func installCmd(streams *cli.IOStreams, cmd *cobra.Command) error {
 
+	// only support Linux at the moment
+	nonRoot, _ := cmd.Flags().GetBool(flagInstallNonRoot)
+	if runtime.GOOS != "linux" {
```
shouldn't this be `nonRoot && runtime.GOOS != "linux"`

blakerouse replied:
Yes it should. Fixed.

blakerouse commented:
> I built Agent from this PR's branch and tried installing it with `--non-root`. I think there are some issues with the socket creation.
> 
> ```
> $ sudo ./elastic-agent install -n --non-root
> Installing in non-interactive mode.
> [    ] Done  [16s]
> Elastic Agent has been successfully installed.
> sudo elastic-agent version
> Binary: 8.12.0-SNAPSHOT (build: 28faa54f1f3c0cdd50c280d8f3a5bad6a348046a at 2023-11-08 06:54:37 +0000 UTC)
> Daemon: <failed to communicate>
> could not get version. failed to communicate with running daemon: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /run/elastic-agent.sock: connect: no such file or directory"
> Use --binary-only flag to skip trying to retrieve version from running daemon
> ```

You didn't read the description of the PR ;-). You cannot call `sudo elastic-agent version` that is running it as root, you installed it as non-root. See "How to test this PR locally" of the PR description.

ycombinator commented:
> > I built Agent from this PR's branch and tried installing it with `--non-root`. I think there are some issues with the socket creation.
> > ```
> > $ sudo ./elastic-agent install -n --non-root
> > Installing in non-interactive mode.
> > [    ] Done  [16s]
> > Elastic Agent has been successfully installed.
> > sudo elastic-agent version
> > Binary: 8.12.0-SNAPSHOT (build: 28faa54f1f3c0cdd50c280d8f3a5bad6a348046a at 2023-11-08 06:54:37 +0000 UTC)
> > Daemon: <failed to communicate>
> > could not get version. failed to communicate with running daemon: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial unix /run/elastic-agent.sock: connect: no such file or directory"
> > Use --binary-only flag to skip trying to retrieve version from running daemon
> > ```
> 
> You didn't read the description of the PR ;-). You cannot call `sudo elastic-agent version` that is running it as root, you installed it as non-root. See "How to test this PR locally" of the PR description.

Ah, I see that now 🤦 - thanks!

blakerouse commented:
I have updated the PR to switch to `--unprivileged` and to switch to a structure for passing around the UID/GID between the functions. The structure switched allowed it to be much simpler being that I could define different structures between unix and Windows. I fixed and cleaned up all review comments, this is ready for more reviews, thanks!

pchila commented on pkg/utils/perm_windows.go:
```diff
@@ -6,6 +6,21 @@
+
+// CurrentFileOwner returns the executing UID and GID of the current process.
+func CurrentFileOwner() FileOwner {
+	// TODO(blakerouse): Make this return the current user and group on Windows.
```
Is this TODO to be implemented in this PR ?

blakerouse replied:
It is for a follow up that I am working on, which enabled non Administrator on Windows.

pchila commented on pkg/utils/perm_unix.go:
```diff
@@ -11,6 +11,20 @@ import (
 	"os"
 )
 
+// FileOwner is the ownership a file should have.
```
This describes a user along with its default group, so maybe the name should be `UserGroupIDs` ?

blakerouse replied:
A file is always owned by a user and a group. To me the name is correct, and you cannot have a file owned by multiple groups at the same time.

ycombinator commented on internal/pkg/agent/cmd/install.go:
```diff
@@ -43,6 +47,7 @@ would like the Agent to operate.
 	cmd.Flags().BoolP("force", "f", false, "Force overwrite the current installation and do not prompt for confirmation")
 	cmd.Flags().BoolP("non-interactive", "n", false, "Install Elastic Agent in non-interactive mode which will not prompt on missing parameters but fails instead.")
 	cmd.Flags().String(flagInstallBasePath, paths.DefaultBasePath, "The path where the Elastic Agent will be installed. It must be an absolute path.")
+	cmd.Flags().Bool(flagInstallUnprivileged, false, "Installed Elastic Agent will create an 'elastic-agent' user and run as that user, instead of running as root.")
```
Since we've replaced `--non-root` with `--unprivileged`, might as well remove the `root` reference from this help text as well.

```suggestion
	cmd.Flags().Bool(flagInstallUnprivileged, false, "Installed Elastic Agent will create an 'elastic-agent' user and run as that user, instead of running as the privileged user.")
```

michalpristas replied:
agree, root is OS specific

blakerouse replied:
Updated.

ycombinator reviewed (approved):
LGTM — nice work!

michalpristas commented on internal/pkg/agent/application/info/state_windows.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+func fixInstallMarkerPermissions(markerFilePath string, ownership utils.FileOwner) error {
+	// TODO(blakerouse): Fix the market permissions on Windows.
```
missed or future work? can we link an issue if so?

blakerouse replied:
don't really have a public issue, so I didn't link it

I am working on Windows next, this will not live long.

michalpristas commented on internal/pkg/agent/install/uninstall.go:
```diff
@@ -35,7 +35,8 @@ import (
 	// uninstall the current service
-	svc, err := newService(topPath)
+	// not creating the service, so no need to set the username and group to any value
+	svc, err := newService(topPath, "", "")
```
a lot of these with `""` as args while uid and gid are actually used in not that many use cases. maybe we should have `newService(path)` and `newServiceWithOwner(path, uid,gid)` or fluent alternative `newService(path).WithOwner(uid, gid)`

blakerouse replied:
I cleaned this up to use the options pattern so it doesn't need to be changed everywhere.

michalpristas commented on pkg/utils/perm_windows.go:
```diff
@@ -6,6 +6,21 @@
+	// TODO(blakerouse): Make this return the current user and group on Windows.
+	return FileOwner{
+		UID: "S-1-5-32-544", // Administrator
+		GID: "S-1-5-18",     // SYSTEM
```
may be worth extracting to consts and using these consts in `internal/pkg/agent/install/perms_windows.go` as well. helps tracking usage

blakerouse replied:
Unified this in the PR to use a constant here and anywhere else that used those same strings. It was in a few places before this change.

michalpristas reviewed (approved):
in general it looks good, still few comments, mostly minor. I;d like to see craigs suggested refactor in though
I havent tested this on windows

cmacknz reviewed (approved):
Looks good, I created https://github.com/elastic/elastic-agent/issues/3736 to follow up on ensuring we test this thoroughly.

IMO we should prioritize getting solid tests in place before supporting the other OSes.

My biggest problem so far is reliably typing unprivileged correctly. For whatever reason I always type it as unpriveledged or something.

blakerouse commented:
@cmacknz I cannot merge it because of Sonar and the failed coverage merge in buildkite.

cmacknz commented:
Yeah looks like https://buildkite.com/elastic/elastic-agent/builds/4755#018bb508-189b-47b2-b5b1-edeb428e28e8 actually passed but isn't posting results.

cmacknz commented:
Let's run it one more time and to give Anderson and Tiago time to respond to some of the new comments, then I can force merge it.

cmacknz commented:
buildkite test it

amitkanfer commented:
related: https://github.com/elastic/ingest-dev/issues/1766
